File: 2016/python/day6.py
from collections import Counter
from typing import List


def partI(lines: List[str]) -> str:
    """
    Function to compute the most repeated character in each position
    """

    # computing the number of available position
    m = len(lines[0])

    # found characters
    chars = []

    for pos in range(m):
        c = Counter([line[pos] for line in lines])
        chars.append(c.most_common(1)[0][0])
    return "".join(chars)


def partII(lines: List[str]) -> str:
    """
    Function to compute the second part of the answer
    """

    # computing the number of available position
    m = len(lines[0])

    # found characters
    chars = []

    for pos in range(m):
        c = Counter([line[pos] for line in lines])
        chars.append(c.most_common()[-1][0])
    return "".join(chars)


if __name__ == "__main__":
    # reading the parts
    with open("../input/day6") as f:
        lines = [line.strip() for line in f.readlines()]

    # computing the first answer
    answer1 = partI(lines)
    print(answer1)

    # compute the second answer
    answer2 = partII(lines)
    print(answer2)

File: 2016/rust/src/day6.rs
use counter::Counter;
use std::fs;

// simple parser to parse a single tag
//
fn most_repated_char(signals: &Vec<String>, i: usize) -> char {
    // get the most repeated character in a given position from
    // the vector of strings
    let counts = signals
        .iter()
        .map(|line| line.chars().nth(i))
        .collect::<Counter<_>>();
    counts.k_most_common_ordered(1)[0].0.unwrap()
}
fn partI(signals: &str) -> String {
    // function to get th
    let signals: Vec<String> = signals.split('\n').map(|x| x.to_string()).collect();
    let n = signals.len();
    let m = signals[0].len();

    // lets create the message
    let mut decoded = String::new();
    for pos in 0..m {
        decoded.push(most_repated_char(&signals, pos));
    }

    decoded
}

fn main() {
    let signals = fs::read_to_string("input/day6").expect("file not found");

    // computing the firet answer
    let answer1 = partI(&signals);
    println!("answer1 = {answer1}");
}
